File: 프로그래머스/5/81305. 시험장 나누기/시험장 나누기.py
import sys
sys.setrecursionlimit(20000) # 재귀 한도 초과 방지

def solution(k, num, links):
    n = len(num)
    parent = [-1] * n
    for i, (l, r) in enumerate(links):
        if l != -1: parent[l] = i
        if r != -1: parent[r] = i
    
    root = -1
    for i in range(n):
        if parent[i] == -1:
            root = i
            break
    
    def check(limit):
        cnt = 0 # 끊은 간선의 수
        
        def dfs(node):
            nonlocal cnt
            if node == -1: # 자식이 없을 경우 return
                return 0
            
            [left, right] = [dfs(links[node][0]), dfs(links[node][1])]
            
            # 1. 자신 + 양쪽 자식 합이 제한 내인 경우
            if num[node] + left + right <= limit:
                return num[node] + left + right
            
            # 2. 자신 + 작은 쪽 자식 합이 제한 내인 경우 -> 큰 쪽 하나 끊음
            if num[node] + min(left, right) <= limit:
                cnt += 1
                return num[node] + min(left, right)
            
            # 3. 자신만 겨우 제한 내인 경우 -> 양쪽 다 끊음
            cnt += 2
            return num[node]
        
        dfs(root)
        return cnt + 1 <= k # 끊은 간선의 수 + 1이 그룹의 수

    # 이분 탐색 (Parametric Search)
    [low, high] = [max(num), sum(num)]
    answer = high
    
    while low <= high:
        mid = (low + high) // 2
        if check(mid):
            answer = mid
            high = mid - 1
        else:
            low = mid + 1
            
    return answer

File: 프로그래머스/5/81305. 시험장 나누기/시험장 나누기.js
function solution(k, num, links) {
    const parent = Array(num.length).fill(-1);
    for (const [i, [left, right]] of links.entries()) {
        if (left !== -1) parent[left] = i;
        if (right !== -1) parent[right] = i;
    }
    const root = parent.findIndex(n => n === -1);
    
    // 재귀 없이 후위 순회(Post-order) 순서 만들기 (스택 활용)
    const postOrder = [];
    const stack = [root];
    while (stack.length) {
        const node = stack.pop();
        if (node === -1) continue;
        postOrder.push(node);
        stack.push(links[node][0]); // 왼쪽 자식
        stack.push(links[node][1]); // 오른쪽 자식
    }
    postOrder.reverse(); // 자식들이 부모보다 앞에 오도록 뒤집기

    function check(limit) {
        let count = 0;
        
        const currentSums = Array(num.length).fill(0); // 각 노드에서의 서브트리 합 저장

        // 재귀 대신 반복문으로 하위 노드부터 처리
        for (const node of postOrder) {
            const leftSum = links[node][0] === -1 ? 0 : currentSums[links[node][0]];
            const rightSum = links[node][1] === -1 ? 0 : currentSums[links[node][1]];

            if (num[node] + leftSum + rightSum <= limit) {
                currentSums[node] = num[node] + leftSum + rightSum;
            } else if (num[node] + Math.min(leftSum, rightSum) <= limit) {
                count++;
                currentSums[node] = num[node] + Math.min(leftSum, rightSum);
            } else {
                count += 2;
                currentSums[node] = num[node];
            }
        }
        return count + 1 <= k;
    }
    
    let [low, high] = [Math.max(...num), num.reduce((acc, cur) => acc + cur, 0)];
    let answer = high;
    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        if (check(mid)) {
            answer = mid;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }
    return answer;
}